File: ggcommons/messaging/messaging_client.py
import logging
import uuid
from argparse import Namespace
from typing import Callable, List, Optional

from ggcommons.messaging.message import Message
from ggcommons.messaging.messaging_provider import MessagingProvider
from ggcommons.messaging.messaging_config import MessagingConfiguration
from ggcommons.messaging.providers.greengrass.greengrass_ipc import (
    GreengrassIpcProvider,
)
from ggcommons.messaging.providers.mqtt import MqttProvider
from ggcommons.messaging.providers.standalone_provider import StandaloneProvider
from ggcommons.utils.iou import Iou
from awsiot.greengrasscoreipc.model import QOS

logger = logging.getLogger("MessagingClient")


class MessagingClient:
    _messaging_provider: MessagingProvider = None

    @staticmethod
    def init(args: Namespace, standalone_config_path: str = None, receive_own_messages=False) -> MessagingProvider:
        """Initialize messaging client based on mode and configuration."""
        mode = getattr(args, 'mode', None)
        thing_name = getattr(args, 'thing', None)
        
        logger.info(f"Initializing MessagingClient - mode: {mode}, thing_name: {thing_name}, receive_own_messages: {receive_own_messages}")
        
        # Determine messaging mode
        if mode and len(mode) > 0 and mode[0].upper() == 'STANDALONE':
            logger.info(f"Configuring STANDALONE mode with dual broker support - config file: {standalone_config_path}")
            if not standalone_config_path:
                logger.error("STANDALONE mode specified but no config file path provided")
                raise RuntimeError("STANDALONE mode requires standalone config file path")
            
            logger.debug(f"Loading messaging configuration from: {standalone_config_path}")
            messaging_config = MessagingClient._get_messaging_config(standalone_config_path)
            
            logger.info("Creating StandaloneProvider for dual broker messaging")
            MessagingClient._messaging_provider = StandaloneProvider(messaging_config, thing_name)
            logger.info("STANDALONE mode messaging provider initialized successfully")
        else:
            # Default to IPC mode
            logger.info(f"Configuring Greengrass IPC mode - receive_own_messages: {receive_own_messages}")
            MessagingClient._messaging_provider = GreengrassIpcProvider(
                receive_own_messages
            )
            logger.info("Greengrass IPC messaging provider initialized successfully")
        
        if MessagingClient._messaging_provider is None:
            logger.error("Failed to create messaging provider - provider is None")
            raise RuntimeError("Failed to initialize messaging provider")
        
        logger.info(f"MessagingClient initialization completed - provider type: {type(MessagingClient._messaging_provider).__name__}")
        return MessagingClient._messaging_provider
    
    @staticmethod
    def _get_messaging_config(standalone_config_path: str) -> MessagingConfiguration:
        """Get messaging configuration from standalone config file."""
        logger.debug(f"Loading messaging configuration from file: {standalone_config_path}")
        
        try:
            config = MessagingConfiguration.load_from_file(standalone_config_path)
            logger.debug(f"Successfully loaded messaging configuration from {standalone_config_path}")
            
            logger.debug("Validating messaging configuration")
            if not config.validate():
                logger.error(f"Messaging configuration validation failed for file: {standalone_config_path}")
                raise RuntimeError("Invalid messaging configuration")
            
            logger.info(f"Messaging configuration loaded and validated successfully from: {standalone_config_path}")
            return config
            
        except Exception as e:
            logger.error(f"Failed to load messaging configuration from {standalone_config_path}: {e}")
            raise RuntimeError(f"STANDALONE mode requires valid messaging configuration: {e}")

    @staticmethod
    def shutdown():
        MessagingClient._messaging_provider.disconnect()
        MessagingClient._messaging_provider = None

    @staticmethod
    def get_messaging_provider() -> MessagingProvider:
        return MessagingClient._messaging_provider

    @staticmethod
    def publish(topic: str, msg: Message):
        logger.debug(f"Publishing message to topic: {topic}")
        MessagingClient._messaging_provider.publish(topic, msg)

    @staticmethod
    def publish_raw(topic: str, msg: dict):
        MessagingClient._messaging_provider.publish_raw(topic, msg)

    @staticmethod
    def publish_to_iot_core(topic: str, msg: Message, qos: str):
        logger.debug(f"Publishing message to IoT Core topic: {topic}, QoS: {qos}")
        MessagingClient._messaging_provider.publish_to_iot_core(topic, msg, qos)

    @staticmethod
    def publish_to_iot_core_raw(topic: str, msg: dict, qos: str):
        MessagingClient._messaging_provider.publish_to_iot_core_raw(topic, msg, qos)

    @staticmethod
    def subscribe(
        topic: str,
        callback: Callable[[str, Message], None],
        max_concurrency: int = None,
    ):
        logger.debug(f"Subscribing to topic: {topic}, max_concurrency: {max_concurrency}")
        MessagingClient._messaging_provider.subscribe(topic, callback, max_concurrency)

    @staticmethod
    def subscribe_to_iot_core(
        topic: str,
        callback: Callable[[str, Message], None],
        qos: QOS,
        max_concurrency: int = None,
    ):
        logger.debug(f"Subscribing to IoT Core topic: {topic}, QoS: {qos}, max_concurrency: {max_concurrency}")
        MessagingClient._messaging_provider.subscribe_to_iot_core(
            topic, callback, qos, max_concurrency
        )

    @staticmethod
    def unsubscribe(topic: str):
        MessagingClient._messaging_provider.unsubscribe(topic)

    @staticmethod
    def unsubscribe_from_iot_core(topic: str):
        MessagingClient._messaging_provider.unsubscribe_from_iot_core(topic)

    @staticmethod
    def request(topic: str, msg: Message) -> Iou:
        logger.debug(f"Sending request to topic: {topic}")
        return MessagingClient._messaging_provider.request(topic, msg)

    @staticmethod
    def request_from_iot_core(topic: str, msg: Message) -> Iou:
        logger.debug(f"Sending request to IoT Core topic: {topic}")
        return MessagingClient._messaging_provider.request_from_iot_core(topic, msg)

    @staticmethod
    def cancel_request(iou: Iou) -> Iou:
        return MessagingClient._messaging_provider.cancel_request(iou)

    @staticmethod
    def cancel_request_from_iot_core(iou: Iou) -> Iou:
        return MessagingClient._messaging_provider.cancel_request(iou)

    @staticmethod
    def reply(request: Message, reply: Message):
        MessagingClient._messaging_provider.reply(request, reply)

    @staticmethod
    def reply_to_iot_core(request: Message, reply: Message):
        MessagingClient._messaging_provider.reply_to_iot_core(request, reply)

    @staticmethod
    def topic_matches_sub(sub: str, topic: str) -> bool:
        return MessagingProvider.topic_matches_sub(sub, topic)

    @staticmethod
    def get_native_client():
        return MessagingClient._messaging_provider.get_native_client()

File: libs/java/src/main/java/com/aws/proserve/ggcommons/messaging/MessagingClient.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.aws.proserve.ggcommons.messaging;

import com.aws.proserve.ggcommons.ParsedCommandLine;
import com.aws.proserve.ggcommons.messaging.providers.standalone.StandaloneMessagingProvider;
import com.aws.proserve.ggcommons.messaging.providers.greengrass.GreengrassMessagingProvider;
import com.google.gson.JsonObject;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import software.amazon.awssdk.aws.greengrass.model.QOS;

import java.util.function.BiConsumer;

/**
 * A client for handling messaging operations in Greengrass components. This class provides methods for
 * publishing and subscribing to messages, making requests, and handling replies in both local and IoT Core
 * communication contexts.
 */
public class MessagingClient
{
    protected static final Logger LOGGER = LogManager.getLogger(MessagingClient.class);

    /** Default per-subscription queue bound when a caller does not specify one (drop-oldest on overflow). */
    public static final int DEFAULT_MAX_MESSAGES = 10_000;

    private MessagingProvider messagingProvider;

    /**
     * Protected no-arg constructor for testing/subclassing (e.g. mock messaging clients).
     * Leaves the underlying provider null; subclasses are expected to override the messaging methods.
     */
    protected MessagingClient() {
    }

    /**
     * Package-private constructor for builder pattern.
     */
    MessagingClient(ParsedCommandLine cmdLine, boolean receiveOwnMessages) {
        switch (cmdLine.mode) {
            case GREENGRASS:
                LOGGER.info("GREENGRASS mode specified. Using Greengrass IPC.");
                this.messagingProvider = new GreengrassMessagingProvider(receiveOwnMessages);
                break;
            case STANDALONE:
                LOGGER.info("STANDALONE mode specified. Using dual MQTT clients.");
                try {
                    MessagingConfiguration config = MessagingConfiguration.loadFromFile(cmdLine.standaloneConfigPath);
                    this.messagingProvider = new StandaloneMessagingProvider(config, cmdLine.thingName);
                } catch (Exception e) {
                    LOGGER.fatal("Failed to load standalone messaging configuration: {}", e.getMessage());
                    throw new RuntimeException("Failed to load standalone messaging configuration: " + e.getMessage(), e);
                }
                break;
            default:
                LOGGER.fatal("Invalid mode specified: {}", cmdLine.mode);
                throw new RuntimeException("Invalid mode specified: " + cmdLine.mode);
        }
    }

    /**
     * Publishes a message to a specified topic.
     *
     * @param topic The topic to publish the message to
     * @param msg The message to publish
     */
    public void publish(String topic, Message msg)
    {
        messagingProvider.publish(topic, msg);
        LOGGER.debug("Published IPC message on topic '{}': {}", topic, msg.toString());
    }

    /**
     * Publishes a message to AWS IoT Core with specified quality of service.
     *
     * @param topic The IoT Core topic to publish to
     * @param msg The message to publish
     * @param qos The quality of service level for message delivery
     */
    public void publishToIotCore(String topic, Message msg, QOS qos)
    {
        messagingProvider.publishToIoTCore(topic, msg, qos);
        LOGGER.debug("Published IoT Core message on topic '{}': {}", topic, msg.toString());
    }

    /**
     * Publishes a raw JSON object to a topic without wrapping it in a Message.
     *
     * @param topic The topic to publish to
     * @param metricObject The JSON object to publish
     */
    public void publishRaw(String topic, JsonObject metricObject)
    {
        messagingProvider.publishRaw(topic, metricObject);
    }

    /**
     * Publishes a raw JSON object to a topic without wrapping it in a Message.
     *
     * @param topic The topic to publish to
     * @param metricObject The JSON object to publish
     */
    public void publishToIotCoreRaw(String topic, JsonObject metricObject, QOS qos)
    {
        messagingProvider.publishToIoTCoreRaw(topic, metricObject, qos);
    }

    /**
     * Subscribes to messages on a topic with a callback for message handling.
     *
     * @param topicFilter The topic filter to subscribe to
     * @param callback The callback to invoke when messages are received
     */
    public void subscribe(String topicFilter, BiConsumer<String, Message> callback)
    {
        subscribe(topicFilter, callback, -1);
    }

    public void subscribe(String topicFilter, BiConsumer<String, Message> callback, int maxConcurrency)
    {
        subscribe(topicFilter, callback, maxConcurrency, DEFAULT_MAX_MESSAGES);
    }

    /**
     * @param maxMessages per-subscription queue bound; when the buffer is full the oldest pending
     *     message is dropped with a warning (parity with the Rust/TS providers). {@code <= 0} =
     *     unbounded. Omitting it uses {@link #DEFAULT_MAX_MESSAGES}.
     */
    public void subscribe(String topicFilter, BiConsumer<String, Message> callback, int maxConcurrency, int maxMessages)
    {
        messagingProvider.subscribe(topicFilter, callback, maxConcurrency, maxMessages);
        LOGGER.debug("Subscribed to IPC messages on topic filter {}", topicFilter);
    }

    /**
     * Subscribes to messages from IoT Core with specified quality of service.
     *
     * @param topicFilter The topic filter to subscribe to
     * @param callback The callback to invoke when messages are received
     * @param qos The quality of service level for the subscription
     */
    public void subscribeToIoTCore(String topicFilter, BiConsumer<String, Message> callback, QOS qos)
    {
        subscribeToIoTCore(topicFilter, callback, qos, -1);
    }

    public void subscribeToIoTCore(String topicFilter, BiConsumer<String, Message> callback, QOS qos, int maxConcurrency)
    {
        subscribeToIoTCore(topicFilter, callback, qos, maxConcurrency, DEFAULT_MAX_MESSAGES);
    }

    /** @param maxMessages per-subscription queue bound (drop-oldest + warn on overflow); {@code <= 0} = unbounded. */
    public void subscribeToIoTCore(String topicFilter, BiConsumer<String, Message> callback, QOS qos, int maxConcurrency, int maxMessages)
    {
        messagingProvider.subscribeToIoTCore(topicFilter, callback, qos, maxConcurrency, maxMessages);
        LOGGER.debug("Subscribed to IoT Core messages on topic filter {}", topicFilter);
    }

    /**
     * Sends a request message and returns a future for handling the reply.
     *
     * @param topic The topic to send the request to
     * @param request The request message
     * @return A ReplyFuture for handling the response
     */
    public ReplyFuture request(String topic, Message request)
    {
        return messagingProvider.request(topic, request);
    }

    /**
     * Sends a request message to IoT Core and returns a future for handling the reply.
     *
     * @param topic The IoT Core topic to send the request to
     * @param request The request message
     * @return A ReplyFuture for handling the response
     */
    public ReplyFuture requestFromIoTCore(String topic, Message request)
    {
        return messagingProvider.requestFromIoTCore(topic, request);
    }

    /**
     * Cancels a pending request and cleans up associated resources.
     *
     * @param replyFuture The ReplyFuture associated with the request to cancel
     */
    public void cancelRequest(ReplyFuture replyFuture)
    {
        messagingProvider.cancelRequest(replyFuture);
    }

    public void cancelRequestFromIoTCore(ReplyFuture replyFuture)
    {
        messagingProvider.cancelRequestFromIoTCore(replyFuture);
    }

    /**
     * Sends a reply to a received request message.
     *
     * @param request The original request message
     * @param reply The reply message
     */
    public void reply(Message request, Message reply)
    {
        messagingProvider.reply(request, reply);
        LOGGER.debug("Published reply on topic '{}: {}", request.getHeader().getReplyTo(), reply.toString());
    }

    public void replyToIoTCore(Message request, Message reply)
    {
        messagingProvider.replyToIoTCore(request, reply);
    }

    /**
     * Unsubscribes from messages on a topic.
     *
     * @param topicFilter The topic filter to unsubscribe from
     */
    public void unsubscribe(String topicFilter)
    {
        messagingProvider.unsubscribe(topicFilter);
        LOGGER.debug("Unsubscribed to IPC messages on topic filter {}", topicFilter);
    }

    public void unsubscribeFromIoTCore(String topicFilter)
    {
        messagingProvider.unsubscribeFromIoTCore(topicFilter);
        LOGGER.debug("Unsubscribed to IPC messages on topic filter {}", topicFilter);
    }

    /**
     * Checks if a topic matches a topic filter pattern.
     *
     * @param topicFilter The topic filter pattern
     * @param topic The topic to check
     * @return true if the topic matches the filter, false otherwise
     */
    public static boolean topicMatchesFilter(String topicFilter, String topic)
    {
        return MessagingProvider.topicMatchesFilter(topicFilter, topic);
    }

    /**
     * Closes the underlying messaging provider, releasing connections and background threads.
     */
    public void close()
    {
        if (messagingProvider != null)
        {
            messagingProvider.close();
        }
    }

    /**
     * Returns the underlying native local messaging client implementation.
     *
     * @return The native messaging client object
     */
    public Object getNativeLocalClient()
    {
        return messagingProvider.getNativeLocalClient();
    }

    /**
     * Returns the underlying native iot core messaging client implementation.
     *
     * @return The native messaging client object
     */
    public Object getNativeIotCoreClient()
    {
        return messagingProvider.getNativeIotCoreClient();
    }

}